File: go/core/subagent_test.go
package core

import (
	"context"
	"testing"
)

func TestDelegateToolRunsChildAndReturnsReply(t *testing.T) {
	// The child agent answers the delegated subtask.
	childClient := &fakeClient{scripted: []ChatResponse{{Content: "researched: 42"}}}
	child := NewSmoothAgent(childClient, AgentOptions{Instructions: "researcher"})
	researcher := DelegateTool("researcher", "Delegate a research subtask.", child)

	// The parent calls the delegate tool, then wraps up.
	parentClient := &fakeClient{scripted: []ChatResponse{
		{ToolCalls: []ToolCall{{ID: "c1", Name: "researcher", Arguments: `{"task":"find the answer"}`}}},
		{Content: "the answer is 42"},
	}}
	parent := NewSmoothAgent(parentClient, AgentOptions{Tools: []Tool{researcher}})

	result, err := parent.Run(context.Background(), "delegate to the researcher", nil)
	if err != nil {
		t.Fatal(err)
	}
	if result.Text != "the answer is 42" || result.ToolCalls != 1 {
		t.Fatalf("unexpected result: %+v", result)
	}
}

func TestDelegateToolSchemaRequiresTask(t *testing.T) {
	child := NewSmoothAgent(&fakeClient{scripted: []ChatResponse{{Content: "x"}}}, AgentOptions{})
	tool := DelegateTool("helper", "help", child)
	if tool.Name() != "helper" {
		t.Fatalf("name = %q", tool.Name())
	}
	props, ok := tool.Parameters()["properties"].(map[string]any)
	if !ok {
		t.Fatalf("properties missing: %+v", tool.Parameters())
	}
	if _, ok := props["task"]; !ok {
		t.Fatalf("task property missing: %+v", props)
	}
	required, ok := tool.Parameters()["required"].([]string)
	if !ok || len(required) != 1 || required[0] != "task" {
		t.Fatalf("required should be [task]: %+v", tool.Parameters()["required"])
	}
}

File: typescript/core/test/subagent.test.ts
import { describe, expect, it } from 'vitest';
import { ChatClientLike, delegateTool, SmoothAgent } from '../src/agent.js';

function fake(scripted: Array<{ content: string | null; tool_calls?: unknown }>): ChatClientLike {
    let i = 0;
    return {
        chat: {
            completions: {
                create: async () => {
                    const s = scripted[i++];
                    return { choices: [{ message: { content: s.content, tool_calls: s.tool_calls as never } }] };
                },
            },
        },
    };
}

describe('sub-agent delegation', () => {
    it('runs the child agent and returns its reply as the tool result', async () => {
        const child = new SmoothAgent(fake([{ content: 'researched: 42' }]), { instructions: 'researcher' });
        const researcher = delegateTool('researcher', 'Delegate a research subtask.', child);

        const parent = new SmoothAgent(
            fake([
                { content: null, tool_calls: [{ id: 'c1', function: { name: 'researcher', arguments: '{"task":"find the answer"}' } }] },
                { content: 'the answer is 42' },
            ]),
            { tools: [researcher] },
        );

        const result = await parent.run('delegate to the researcher');
        expect(result.text).toBe('the answer is 42');
        expect(result.toolCalls).toBe(1);
    });

    it('exposes a task-requiring schema', () => {
        const child = new SmoothAgent(fake([{ content: 'x' }]), {});
        const tool = delegateTool('helper', 'help', child);
        expect(tool.name).toBe('helper');
        const params = tool.parameters as { properties: Record<string, unknown>; required: string[] };
        expect(params.properties).toHaveProperty('task');
        expect(params.required).toEqual(['task']);
    });
});
